File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/37_sudoku-solver/solutions/cpp/solution.cpp
#include <vector>

class Solution {
public:
    void solveSudoku(std::vector<std::vector<char>>& board) {
        solve(board);
    }
    
private:
    bool solve(std::vector<std::vector<char>>& board) {
        for (int i = 0; i < 9; ++i) {
            for (int j = 0; j < 9; ++j) {
                if (board[i][j] == '.') {
                    for (char num = '1'; num <= '9'; ++num) {
                        if (isValid(board, i, j, num)) {
                            board[i][j] = num;
                            if (solve(board)) {
                                return true;
                            }
                            board[i][j] = '.';
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    
    bool isValid(const std::vector<std::vector<char>>& board, int row, int col, char num) {
        for (int i = 0; i < 9; ++i) {
            if (board[i][col] == num) return false; // Check row
            if (board[row][i] == num) return false; // Check column
            if (board[3 * (row / 3) + i / 3][3 * (col / 3) + i % 3] == num) return false; // Check 3x3 box
        }
        return true;
    }
};

File: DataSets/DataSet-Old/LeetCodeTasksDevelop/LeetCodeTasks/37_sudoku-solver/solutions/python3/solution.py
class Solution:
    def solveSudoku(self, board):
        """
        Do not return anything, modify board in-place instead.
        """
        def is_valid_move(x, y, val):
            # Check if the value is already in the row
            for j in range(9):
                if board[x][j] == val:
                    return False
            # Check if the value is already in the column
            for i in range(9):
                if board[i][y] == val:
                    return False
            # Check if the value is already in the 3x3 grid
            startRow = x - x % 3
            startCol = y - y % 3
            for i in range(3):
                for j in range(3):
                    if board[startRow + i][startCol + j] == val:
                        return False
            return True
        
        def solve():
            for i in range(9):
                for j in range(9):
                    if board[i][j] == '.':
                        for num in '123456789':
                            if is_valid_move(i, j, num):
                                board[i][j] = num
                                if solve():
                                    return True
                                board[i][j] = '.'
                        return False
            return True
        
        solve()
